blogs/2019/11/26/custom-data-format.md

```diff
@@ -0,0 +1,94 @@
+
+To modernize the HTML and CSS language support in Visual Studio Code, we designed the [Custom Data Format](https://github.com/microsoft/vscode-custom-data), a declarative JSON format for describing HTML and CSS entities. VS Code's HTML and CSS language servers can read data defined in this format and provide language support for the newly defined HTML and CSS entities.
+
+In this blog post, we explain the data format and how users and extension authors can take advantage of it.
+
+## Using Custom Data Format
+
+VS Code provides information about HTML/CSS entities in auto completion and hover information:
+
+![HTML and CSS Language Features](html-css-language-features.png)
+
+With the Custom Data Format, you can define new HTML/CSS entities and get auto completion, hover information and other language features. To get started, write a JSON file:
+
+```json
+{
+  "version": 1.1,
+  "tags": [
+    {
+      "name": "my-button",
+      "description": "My button. You should use it as in `<my-button type='alert'></mybutton>`.",
+      "references": [
+        {
+          "name": "Bootstrap buttons",
+          "url": "https://getbootstrap.com/docs/4.0/components/buttons/"
+        }
+      ]
+    }
+  ]
+}
+```
+
+Then configure `html.customData` to point to the data file:
+
+```json
+{
+  "html.customData": ["./html.html-data.json"]
+}
+```
+
+You would get language features for the defined entities:
+
+![Custom Data helloworld](custom-data-helloworld.png)
+
+You can find detailed usage instruction and samples in the [microsoft/vscode-custom-data](https://github.com/microsoft/vscode-custom-data) repository.
+
+## Using Custom Data Format for language servers
+
+The [vscode-html-languageservice](https://github.com/Microsoft/vscode-html-languageservice) and [vscode-css-languageservice](https://github.com/Microsoft/vscode-css-languageservice) libraries allow extension authors to easily implement language servers for languages that extend or embed HTML and CSS. Recently, we added API for loading custom data:
+
+```ts
+import { getLanguageService } from 'vscode-html-languageservice'
+
+getLanguageService({
+  customDataProviders: [...]
+})
+```
+
+By loading different data sets, one can easily create an extension that supports a superset or a subset of HTML/CSS. For example, the [MAVO](https://github.com/octref/vscode-mavo) extension load all HTML entities with MAVO-specific global attributes to get auto completion and hover information for those attributes. It is also possible to write extensions for HTML4, XHTML or SVG by restricitng the entities being loaded.
+
```

Comment:
The above section is best thought of as a behind the scenes topic with a reference to "You can also use this yourself.."